thread_sync.c

```diff
@@ -1430,26 +1430,53 @@ struct rb_condvar {
  *
  *  ConditionVariable objects augment class Mutex. Using condition variables,
  *  it is possible to suspend while in the middle of a critical section until a
- *  resource becomes available.
+ *  condition is met, such as a resource becomes available.
+ *
+ *  Due to non-deterministic scheduling and spurious wake-ups, users of
+ *  condition variables should always use a separate boolean predicate (such as
+ *  reading from a boolean variable) to check if the condition is actually met
+ *  before starting to wait, and should wait in a loop, re-checking the
+ *  condition every time the ConditionVariable is waken up.  The idiomatic way
+ *  of using condition variables is calling the +wait+ method in an +until+
+ *  loop with the predicate as the loop condition.
  *
  *  Example:
  *
  *    mutex = Thread::Mutex.new
- *    resource = Thread::ConditionVariable.new
+ *
+ *    resource_available = false
+ *    condvar = Thread::ConditionVariable.new
  *
  *    a = Thread.new {
- *	 mutex.synchronize {
- *	   # Thread 'a' now needs the resource
- *	   resource.wait(mutex)
- *	   # 'a' can now have the resource
- *	 }
+ *      mutex.synchronize {
+ *        # Thread 'a' now needs the resource.
+ *        # Note:
+ *        # (1) Thread 'b' may be scheduled before thread 'a', and may run so
+ *        #     fast that it have already made the resource available before
+ *        #     thread 'a' start.  Therefore, we should check
+ *        #     'resource_available' before starting to wait.
+ *        # (2) The 'wait' method may spuriously wake up without signalling.
+ *        #     Therefore, we should recheck 'resource_available' after waking
+ *        #     up, and go back to wait if the condition is not actually met.
+ *        # (3) As soon as the resource is available, a third thread (not
+ *        #     necessarily waiting on 'condvar') may be scheduled first and
+ *        #     may have acquired the 'mutex' and consumed the resource
+ *        #     before thread 'a' wakes up.  This implies rechecking, too.
+ *        condvar.wait(mutex) until resource_available
+ *
+ *        # 'a' can now have the resource
+ *      }
  *    }
  *
  *    b = Thread.new {
- *	 mutex.synchronize {
- *	   # Thread 'b' has finished using the resource
- *	   resource.signal
- *	 }
+ *      mutex.synchronize {
+ *        # Thread 'b' makes the resource available.
+ *        resource_available = true
+ *
+ *        # In case any thread is waiting on the condition variable, we notify
+ *        # one thread about the availability of the resource.
+ *        condvar.signal
+ *      }
  *    }
```

Comment:
I've noticed that this example is incorrect because we never call `a.join` or `b.join`, so the main thread terminates before the other two threads are able to do anything.

Reply:
Well, the example is still correct if it is part of a larger program where the main thread runs almost forever.  But I can still add `a.join` and `b.join` to the end so that the user can copy and paste this example to a script and run it.